src/cose/mac0.rs

```diff
@@ -0,0 +1,466 @@
+}
+
+/// Result with error type: [`Error`].
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// Result for verification of a COSE_Mac0.
+#[derive(Debug)]
+pub enum VerificationResult {
+    Success,
+    Failure(String),
+    Error(Error),
+}
+
+impl VerificationResult {
+    /// Result of verification.
+    ///
+    /// False implies the signature is inauthentic or the verification algorithm encountered an
+    /// error.
+    pub fn success(&self) -> bool {
+        matches!(self, VerificationResult::Success)
+    }
+
+    /// Translate to a std::result::Result.
+    ///
+    /// Converts failure reasons and errors into a String.
+    pub fn to_result(self) -> Result<(), String> {
+        match self {
+            VerificationResult::Success => Ok(()),
+            VerificationResult::Failure(reason) => Err(reason),
+            VerificationResult::Error(e) => Err(format!("{}", e)),
+        }
+    }
+
+    /// Retrieve the error if the verification algorithm encountered an error.
+    pub fn to_error(self) -> Option<Error> {
+        match self {
+            VerificationResult::Error(e) => Some(e),
+            _ => None,
+        }
+    }
+}
+
+impl PreparedCoseMac0 {
+    pub fn new(
+        builder: coset::CoseMac0Builder,
+        detached_payload: Option<Vec<u8>>,
+        aad: Option<Vec<u8>>,
+        tagged: bool,
+    ) -> Result<Self> {
+        let cose_mac0 = builder.build();
+
+        // Check if the payload is present and if it is attached or detached.
+        // Needs to be exclusively attached or detached.
+        let payload = match (cose_mac0.payload.as_ref(), detached_payload.as_ref()) {
+            (Some(_), Some(_)) => return Err(Error::DoublePayload),
+            (None, None) => return Err(Error::NoPayload),
+            (Some(payload), None) => payload,
+            (None, Some(payload)) => payload,
+        };
+        // Create the signature payload ot be used later on signing.
+        let tag_payload = mac_structure_data(
+            MacContext::CoseMac0,
+            cose_mac0.protected.clone(),
+            aad.unwrap_or_default().as_ref(),
+            payload,
+        );
+
+        Ok(Self {
+            tagged,
+            cose_mac0: CoseMac0 {
+                tagged,
+                inner: cose_mac0,
+            },
+            tag_payload,
+        })
+    }
+
+    /// Returns the signature payload that needs to be used to tag it.
+    pub fn signature_payload(&self) -> &[u8] {
+        &self.tag_payload
+    }
+
+    /// Finalize the COSE_Mac0 by adding the tag.
+    pub fn finalize(self, tag: Vec<u8>) -> CoseMac0 {
+        let mut cose_mac0 = self.cose_mac0;
+        cose_mac0.inner.tag = tag;
+        cose_mac0
+    }
+}
+
+impl CoseMac0 {
+    /// Verify that the tag of a `COSE_Mac0` is authentic.
+    pub fn verify(
+        &self,
+        verifier: &Hmac<Sha256>,
+        detached_payload: Option<Vec<u8>>,
+        external_aad: Option<Vec<u8>>,
```

Comment:
Again `detached_payload` and `external_aad` are never consumed and shoud be passed by reference to avoid unnecessary clones when the function is called.